File: ios/MenuMakerUITests/PageObjects/FavoritesPage.swift
//
//  FavoritesPage.swift
//  MenuMakerUITests
//
//  Page Object for Favorites/Bookmarks Screen
//

import XCTest

struct FavoritesPage {
    let app: XCUIApplication

    // MARK: - Elements

    var favoritesList: XCUIElementQuery {
        app.scrollViews.otherElements.matching(identifier: "FavoriteItem")
    }

    var firstFavorite: XCUIElement {
        favoritesList.firstMatch
    }

    var emptyStateMessage: XCUIElement {
        app.staticTexts.matching(NSPredicate(format: "label CONTAINS[c] 'no favorite' OR label CONTAINS[c] 'add some'")).firstMatch
    }

    var favoriteButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS '❤' OR label CONTAINS[c] 'favorite' OR identifier CONTAINS 'favorite'")).firstMatch
    }

    var removeButtons: XCUIElementQuery {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'remove' OR label CONTAINS[c] 'delete'"))
    }

    var exploreButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'explore' OR label CONTAINS[c] 'browse'")).firstMatch
    }

    var sortButton: XCUIElement {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'sort'")).firstMatch
    }

    var filterButtons: XCUIElementQuery {
        app.buttons.matching(NSPredicate(format: "label CONTAINS[c] 'all' OR label CONTAINS[c] 'sellers' OR label CONTAINS[c] 'items'"))
    }

    var searchBar: XCUIElement {
        app.searchFields.firstMatch
    }

    // MARK: - Actions

    @discardableResult
    func tapFirstFavorite() -> FavoritesPage {
        firstFavorite.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func tapFavorite(at index: Int) -> FavoritesPage {
        let favorite = favoritesList.element(boundBy: index)
        if favorite.waitForExistence(timeout: 2) {
            favorite.tap()
            sleep(1)
        }
        return self
    }

    @discardableResult
    func removeFavorite(at index: Int = 0) -> FavoritesPage {
        let favorite = favoritesList.element(boundBy: index)
        favorite.swipeLeft()

        let removeButton = app.buttons["Delete"]
        if removeButton.waitForExistence(timeout: 1) {
            removeButton.tap()
            sleep(1)
        }

        return self
    }

    @discardableResult
    func tapFavoriteButton() -> FavoritesPage {
        favoriteButton.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func tapExplore() -> FavoritesPage {
        exploreButton.tap()
        sleep(1)
        return self
    }

    @discardableResult
    func searchFavorites(_ query: String) -> FavoritesPage {
        if searchBar.waitForExistence(timeout: 1) {
            searchBar.tap()
            searchBar.typeText(query)
            sleep(1)
        }
        return self
    }

    @discardableResult
    func pullToRefresh() -> FavoritesPage {
        let scrollView = app.scrollViews.firstMatch
        if scrollView.exists {
            let start = scrollView.coordinate(withNormalizedOffset: CGVector(dx: 0.5, dy: 0.2))
            let end = scrollView.coordinate(withNormalizedOffset: CGVector(dx: 0.5, dy: 0.8))
            start.press(forDuration: 0, thenDragTo: end)
            sleep(1)
        }
        return self
    }

    // MARK: - Assertions

    @discardableResult
    func assertScreenDisplayed(timeout: TimeInterval = 2) -> FavoritesPage {
        XCTAssertTrue(firstFavorite.waitForExistence(timeout: timeout) ||
                     emptyStateMessage.waitForExistence(timeout: timeout),
                     "Favorites screen should be displayed")
        return self
    }

    @discardableResult
    func assertFavoritesDisplayed() -> FavoritesPage {
        XCTAssertTrue(firstFavorite.exists, "Favorites should be displayed")
        return self
    }

    @discardableResult
    func assertEmptyState() -> FavoritesPage {
        XCTAssertTrue(emptyStateMessage.exists, "Empty state should be displayed")
        return self
    }

    @discardableResult
    func assertFavoriteCount(_ expectedCount: Int) -> FavoritesPage {
        let actualCount = favoritesList.count
        XCTAssertEqual(actualCount, expectedCount,
                      "Should have \(expectedCount) favorites, found \(actualCount)")
        return self
    }

    @discardableResult
    func assertFavoriteAdded(_ name: String) -> FavoritesPage {
        let favoriteLabel = app.staticTexts[name]
        XCTAssertTrue(favoriteLabel.exists, "Favorite '\(name)' should be added")
        return self
    }

    @discardableResult
    func assertFavoriteRemoved(_ name: String) -> FavoritesPage {
        let favoriteLabel = app.staticTexts[name]
        XCTAssertFalse(favoriteLabel.exists, "Favorite '\(name)' should be removed")
        return self
    }
}

File: android/app/src/androidTest/kotlin/com/menumaker/pageobjects/FavoritesPage.kt
package com.menumaker.pageobjects

import androidx.compose.ui.test.*
import androidx.compose.ui.test.junit4.ComposeTestRule

/**
 * Page Object for Favorites/Bookmarks Screen
 * Provides fluent API for favorites interactions
 */
class FavoritesPage(private val composeTestRule: ComposeTestRule) {

    // Elements
    private val favoritesList = composeTestRule.onAllNodesWithTag("FavoriteItem")
    private val emptyStateMessage = composeTestRule.onNode(
        hasText("no favorite", substring = true, ignoreCase = true) or
        hasText("add some", substring = true, ignoreCase = true)
    )
    private val favoriteButton = composeTestRule.onNodeWithTag("favorite-button")
    private val exploreButton = composeTestRule.onNode(
        hasText("explore", substring = true, ignoreCase = true) or
        hasText("browse", substring = true, ignoreCase = true)
    )
    private val sortButton = composeTestRule.onNodeWithText("sort", substring = true, ignoreCase = true)
    private val searchBar = composeTestRule.onNodeWithTag("search-bar")

    // Actions
    fun tapFirstFavorite(): FavoritesPage {
        if (favoritesList.fetchSemanticsNodes().isNotEmpty()) {
            favoritesList[0].performClick()
            Thread.sleep(1000)
        }
        return this
    }

    fun tapFavorite(index: Int): FavoritesPage {
        if (favoritesList.fetchSemanticsNodes().size > index) {
            favoritesList[index].performClick()
            Thread.sleep(1000)
        }
        return this
    }

    fun removeFavorite(index: Int = 0): FavoritesPage {
        if (favoritesList.fetchSemanticsNodes().size > index) {
            favoritesList[index].performTouchInput {
                swipeLeft()
            }
            // Tap delete button if it appears
            val deleteButton = composeTestRule.onNodeWithText("Delete", ignoreCase = true)
            try {
                deleteButton.performClick()
                Thread.sleep(1000)
            } catch (e: AssertionError) {
                // Delete button not found
            }
        }
        return this
    }

    fun tapFavoriteButton(): FavoritesPage {
        favoriteButton.performClick()
        Thread.sleep(1000)
        return this
    }

    fun tapExplore(): FavoritesPage {
        exploreButton.performClick()
        Thread.sleep(1000)
        return this
    }

    fun searchFavorites(query: String): FavoritesPage {
        searchBar.performTextInput(query)
        Thread.sleep(1000)
        return this
    }

    // Assertions
    fun assertScreenDisplayed(): FavoritesPage {
        composeTestRule.waitUntil(timeoutMillis = 2000) {
            favoritesList.fetchSemanticsNodes().isNotEmpty() ||
            try { emptyStateMessage.assertExists(); true } catch (e: AssertionError) { false }
        }
        return this
    }

    fun assertFavoritesDisplayed(): FavoritesPage {
        assert(favoritesList.fetchSemanticsNodes().isNotEmpty()) {
            "Favorites should be displayed"
        }
        return this
    }

    fun assertEmptyState(): FavoritesPage {
        emptyStateMessage.assertExists()
        return this
    }

    fun assertFavoriteCount(expectedCount: Int): FavoritesPage {
        val actualCount = favoritesList.fetchSemanticsNodes().size
        assert(actualCount == expectedCount) {
            "Should have $expectedCount favorites, found $actualCount"
        }
        return this
    }

    fun assertFavoriteAdded(name: String): FavoritesPage {
        composeTestRule.onNodeWithText(name).assertExists()
        return this
    }

    fun assertFavoriteRemoved(name: String): FavoritesPage {
        composeTestRule.onNodeWithText(name).assertDoesNotExist()
        return this
    }
}
